python-package/lightgbm/callback.py

```diff
@@ -231,124 +231,135 @@ def early_stopping(stopping_rounds: int, first_metric_only: bool = False, verbos
-        enabled = not any(env.params.get(boost_alias, "") == 'dart' for boost_alias
-                          in _ConfigAliases.get("boosting"))
-        if not enabled:
-            _log_warning('Early stopping is not available in dart mode')
-            return
-        if not env.evaluation_result_list:
-            raise ValueError('For early stopping, '
-                             'at least one dataset and eval metric is required for evaluation')
-
-        if stopping_rounds <= 0:
-            raise ValueError("stopping_rounds should be greater than zero.")
-
-        if verbose:
-            _log_info(f"Training until validation scores don't improve for {stopping_rounds} rounds")
-
-        # reset storages
-        best_score = []
-        best_iter = []
-        best_score_list = []
-        cmp_op = []
-        first_metric = ''
-
-        n_metrics = len(set(m[1] for m in env.evaluation_result_list))
-        n_datasets = len(env.evaluation_result_list) // n_metrics
-        if isinstance(min_delta, list):
-            if not all(t >= 0 for t in min_delta):
-                raise ValueError('Values for early stopping min_delta must be non-negative.')
-            if len(min_delta) == 0:
-                if verbose:
-                    _log_info('Disabling min_delta for early stopping.')
-                deltas = [0.0] * n_datasets * n_metrics
-            elif len(min_delta) == 1:
-                if verbose:
-                    _log_info(f'Using {min_delta[0]} as min_delta for all metrics.')
-                deltas = min_delta * n_datasets * n_metrics
-            else:
-                if len(min_delta) != n_metrics:
-                    raise ValueError('Must provide a single value for min_delta or as many as metrics.')
-                if first_metric_only and verbose:
-                    _log_info(f'Using only {min_delta[0]} as early stopping min_delta.')
-                deltas = min_delta * n_datasets
-        else:
-            if min_delta < 0:
-                raise ValueError('Early stopping min_delta must be non-negative.')
-            if min_delta > 0 and n_metrics > 1 and not first_metric_only and verbose:
-                _log_info(f'Using {min_delta} as min_delta for all metrics.')
-            deltas = [min_delta] * n_datasets * n_metrics
-
-        # split is needed for "<dataset type> <metric>" case (e.g. "train l1")
-        first_metric = env.evaluation_result_list[0][1].split(" ")[-1]
-        for eval_ret, delta in zip(env.evaluation_result_list, deltas):
-            best_iter.append(0)
-            best_score_list.append(None)
-            if eval_ret[3]:  # greater is better
-                best_score.append(float('-inf'))
-                cmp_op.append(partial(_gt_delta, delta=delta))
+    class _early_stopping_callback:
+        def __init__(self, stopping_rounds: int, first_metric_only: bool = False, verbose: bool = True, min_delta: Union[float, List[float]] = 0.0) -> None:
```

Comment:
```suggestion
        def __init__(
            self,
            stopping_rounds: int,
            first_metric_only: bool = False,
            verbose: bool = True,
            min_delta: Union[float, List[float]] = 0.0
        ) -> None:
```

Thanks for adding type hints here! Could you please use this suggested indentation? We've slowly been adopting type hints in this package (see the discussion and PRs linked to #3756), and have used the convention of this one-argument-per-line formatting when there is more than one argument.

For example:

https://github.com/microsoft/LightGBM/blob/717631af1836be9b03444ece485f3a3b5dbae5da/python-package/lightgbm/dask.py#L167-L177